internal/datastore/proxy/schemacaching/watchingcache.go

```diff
@@ -0,0 +1,397 @@
+	caveatCache    *schemaWatchCache[*core.CaveatDefinition]
+}
+
+// startWatchingCacheProxy starts a watching cache proxy's watch goroutine and returns the proxy.
+func startWatchingCacheProxy(wrapped datastore.Datastore, watchable datastore.SchemaWatchableDatastore, gcWindow time.Duration) *watchingCachingProxy {
+	proxy := &watchingCachingProxy{
+		Datastore: wrapped,
+
+		gcWindow: gcWindow,
+		closed:   make(chan bool, 2),
+
+		namespaceCache: newSchemaWatchCache[*core.NamespaceDefinition](
+			datastore.NewNamespaceNotFoundErr,
+			func(ctx context.Context, name string, revision datastore.Revision) (*core.NamespaceDefinition, datastore.Revision, error) {
+				return wrapped.SnapshotReader(revision).ReadNamespaceByName(ctx, name)
+			},
+			func(ctx context.Context, names []string, revision datastore.Revision) ([]datastore.RevisionedDefinition[*core.NamespaceDefinition], error) {
+				return wrapped.SnapshotReader(revision).LookupNamespacesWithNames(ctx, names)
+			},
+		),
+		caveatCache: newSchemaWatchCache[*core.CaveatDefinition](
+			datastore.NewCaveatNameNotFoundErr,
+			func(ctx context.Context, name string, revision datastore.Revision) (*core.CaveatDefinition, datastore.Revision, error) {
+				return wrapped.SnapshotReader(revision).ReadCaveatByName(ctx, name)
+			},
+			func(ctx context.Context, names []string, revision datastore.Revision) ([]datastore.RevisionedDefinition[*core.CaveatDefinition], error) {
+				return wrapped.SnapshotReader(revision).LookupCaveatsWithNames(ctx, names)
+			},
+		),
+	}
+	proxy.startWatch(watchable)
+	return proxy
+}
+
+func (p *watchingCachingProxy) SnapshotReader(rev datastore.Revision) datastore.Reader {
+	delegateReader := p.Datastore.SnapshotReader(rev)
+	return &watchingCachingReader{delegateReader, rev, p}
+}
+
+func (p *watchingCachingProxy) ReadWriteTx(
+	ctx context.Context,
+	f datastore.TxUserFunc,
+	opts ...options.RWTOptionsOption,
+) (datastore.Revision, error) {
+	return p.Datastore.ReadWriteTx(ctx, func(delegateRWT datastore.ReadWriteTransaction) error {
+		// NOTE: we always use the standard approach cache here, as it stores changes within the transaction
+		// itself, and should not impact the overall updating cache.
+		rwt := &definitionCachingRWT{delegateRWT, &sync.Map{}}
+		return f(rwt)
+	}, opts...)
+}
+
+func (p *watchingCachingProxy) startWatch(watchable datastore.SchemaWatchableDatastore) {
+	headRev, err := watchable.HeadRevision(context.Background())
+	if err != nil {
+		p.namespaceCache.setFallbackMode()
+		p.caveatCache.setFallbackMode()
+		log.Warn().Err(err).Msg("received error in schema watch")
+		return
+	}
+
+	p.namespaceCache.setCheckpointRevision(headRev)
+	p.caveatCache.setCheckpointRevision(headRev)
+
+	// Start watching for expired entries to be GCed.
+	go (func() {
+		for {
+			select {
+			case <-p.closed:
+				log.Debug().Msg("GC routine for watch closed")
+				return
+
+			case <-time.After(time.Hour):
```

Comment:
should this be configurable?

Reply:
I can certainly add an option for it, if you think its important to do so